ccl/multiregionccl: TestRegionLivenessProber failed

ccl/multiregionccl.TestRegionLivenessProber [failed](https://tanzanite.cluster.engflow.com/invocations/default/628f4b1e-e8b7-41f6-a501-bf15569cbdae?testReportRun=16&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [1f53a61732c6f9e48622f45d276eb5e1c343ad30](https://github.com/cockroachdb/cockroach/commits/1f53a61732c6f9e48622f45d276eb5e1c343ad30):


```
=== RUN   TestRegionLivenessProber
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRegionLivenessProber1923685213
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:93: (TestRegionLivenessProber)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:93: (TestRegionLivenessProber)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:93: (TestRegionLivenessProber)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    regionliveness_test.go:115: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:115
        	Error:      	Received unexpected error:
        	            	pq: count-lease timed out reading from a region: operation "count-leases-region" timed out after 501ms (given timeout 500ms): count-leases: failed to read query result: query execution canceled
        	Test:       	TestRegionLivenessProber
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestRegionLivenessProber1923685213
--- FAIL: TestRegionLivenessProber (69.09s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=16</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionLivenessProber.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34695

ccl/multiregionccl.TestRegionLivenessProber [failed](https://tanzanite.cluster.engflow.com/invocations/default/a55a8dce-42ad-45bf-b129-d66c8b9782d2?testReportRun=6&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [01c28b7dbc3177361e302525703fb73bd0d6fc8b](https://github.com/cockroachdb/cockroach/commits/01c28b7dbc3177361e302525703fb73bd0d6fc8b):


```
=== RUN   TestRegionLivenessProber
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestRegionLivenessProber3069916602
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:93: (TestRegionLivenessProber)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:93: (TestRegionLivenessProber)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization

pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:93: (TestRegionLivenessProber)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    regionliveness_test.go:115: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl_test/pkg/ccl/multiregionccl/regionliveness_test.go:115
        	Error:      	Received unexpected error:
        	            	pq: count-lease timed out reading from a region: operation "count-leases-region" timed out after 501ms (given timeout 500ms): count-leases: failed to read query result: query execution canceled
        	Test:       	TestRegionLivenessProber
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestRegionLivenessProber3069916602
--- FAIL: TestRegionLivenessProber (65.85s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRegionLivenessProber.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/multiregionccl/regionliveness_test.go
+++ b/pkg/ccl/multiregionccl/regionliveness_test.go
@@ -43,6 +43,10 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+const (
+	testingRegionLivenessProbeTimeout = time.Second * 2
+)
+
 func TestRegionLivenessProber(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -84,7 +88,6 @@ func TestRegionLivenessProber(t *testing.T) {
 	var tenants []serverutils.ApplicationLayerInterface
 	var tenantSQL []*gosql.DB
 	blockProbeQuery := atomic.Bool{}
-	defer regionliveness.TestingSetProbeLivenessTimeout(500 * time.Millisecond)()
 
 	for _, s := range testCluster.Servers {
 		tenantArgs := base.TestTenantArgs{
@@ -97,7 +100,7 @@ func TestRegionLivenessProber(t *testing.T) {
 						const probeQuery = "SELECT count(*) FROM system.sql_instances WHERE crdb_region = $1::system.crdb_internal_region"
 						if strings.Contains(stmt, probeQuery) && blockProbeQuery.Swap(false) {
 							// Timeout this query intentionally.
-							time.Sleep(1 * time.Second)
+							time.Sleep(testingRegionLivenessProbeTimeout + time.Second)
 						}
 					},
 				},
@@ -114,6 +117,11 @@ func TestRegionLivenessProber(t *testing.T) {
 		_, err = tenantSQL[0].Exec(fmt.Sprintf("ALTER DATABASE system ADD REGION '%s'", expectedRegions[i]))
 		require.NoError(t, err)
 	}
+	// Override the table timeout probe for testing.
+	for _, ts := range tenants {
+		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeout)
+	}
+
 	var cachedRegionProvider *regions.CachedDatabaseRegions
 	cachedRegionProvider, err = regions.NewCachedDatabaseRegions(ctx, tenants[0].DB(), tenants[0].LeaseManager().(*lease.Manager))
 	require.NoError(t, err)
@@ -214,8 +222,6 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 		"us-west",
 	}
 	detectLeaseWait := atomic.Bool{}
-	defer regionliveness.TestingSetProbeLivenessTimeout(1 * time.Second)()
-
 	testCluster, _, cleanup := multiregionccltestutils.TestingCreateMultiRegionClusterWithRegionList(t,
 		expectedRegions,
 		1,
@@ -248,9 +254,9 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 							if targetCount.Add(1) != 1 {
 								return
 							}
-							time.Sleep(time.Second * 2)
+							time.Sleep(testingRegionLivenessProbeTimeout + time.Second)
 						} else if strings.Contains(stmt, probeQuery) {
-							time.Sleep(time.Second * 2)
+							time.Sleep(testingRegionLivenessProbeTimeout + time.Second)
 							targetCount.Swap(0)
 							detectLeaseWait.Swap(false)
 						}
@@ -271,6 +277,10 @@ func TestRegionLivenessProberForLeases(t *testing.T) {
 	}
 	_, err = tenantSQL[0].Exec("ALTER DATABASE system SURVIVE ZONE FAILURE")
 	require.NoError(t, err)
+	// Override the table timeout probe for testing.
+	for _, ts := range tenants {
+		regionliveness.RegionLivenessProbeTimeout.Override(ctx, &ts.ClusterSettings().SV, testingRegionLivenessProbeTimeout)
+	}
 	// Create a new table and have it used on all nodes.
 	_, err = tenantSQL[0].Exec("CREATE TABLE t1(j int)")
 	require.NoError(t, err)

--- a/pkg/sql/catalog/lease/count.go
+++ b/pkg/sql/catalog/lease/count.go
@@ -127,7 +127,7 @@ func countLeasesByRegion(
 			return err
 		}
 		var err error
-		if hasTimeout, timeout := prober.GetTableTimeout(); hasTimeout {
+		if hasTimeout, timeout := prober.GetProbeTimeout(); hasTimeout {
 			err = timeutil.RunWithTimeout(ctx, "count-leases-region", timeout, queryRegionRows)
 		} else {
 			err = queryRegionRows(ctx)

--- a/pkg/sql/regionliveness/prober.go
+++ b/pkg/sql/regionliveness/prober.go
@@ -34,8 +34,15 @@ import (
 
 var RegionLivenessEnabled = settings.RegisterBoolSetting(settings.ApplicationLevel,
 	"sql.region_liveness.enabled",
-	"enables region liveness for system databases (experimental)",
+	"enables region liveness for system databases",
 	false, /* disabled */
+	settings.WithVisibility(settings.Reserved))
+
+var RegionLivenessProbeTimeout = settings.RegisterDurationSetting(settings.ApplicationLevel,
+	"sql.region_liveness.probe.timeout",
+	"set the probing timeout for region liveness, which will be the maximum "+
+		"time a query to a region can take before it starts getting quarantined",
+	15*time.Second, /* 15 seconds */
 	settings.WithVisibility(settings.Reserved))
 
 // LiveRegions are regions which are currently still avaialble,
@@ -66,9 +73,9 @@ type Prober interface {
 	// QueryLiveness can be used to get the list of regions which are currently
 	// accessible.
 	QueryLiveness(ctx context.Context, txn *kv.Txn) (LiveRegions, error)
-	// GetTableTimeout gets maximum timeout waiting on a table before issuing
+	// GetProbeTimeout gets maximum timeout waiting on a table before issuing
 	// liveness queries.
-	GetTableTimeout() (bool, time.Duration)
+	GetProbeTimeout() (bool, time.Duration)
 }
 
 // RegionProvider abstracts the lookup of regions (see regions.Provider).
@@ -89,16 +96,6 @@ type livenessProber struct {
 	settings        *clustersettings.Settings
 }
 
-var probeLivenessTimeout = 15 * time.Second
-
-func TestingSetProbeLivenessTimeout(newTimeout time.Duration) func() {
-	oldTimeout := probeLivenessTimeout
-	probeLivenessTimeout = newTimeout
-	return func() {
-		probeLivenessTimeout = oldTimeout
-	}
-}
-
 // NewLivenessProber creates a new region liveness prober.
 func NewLivenessProber(
 	db isql.DB, cachedDBRegions CachedDatabaseRegions, settings *clustersettings.Settings,
@@ -113,13 +110,14 @@ func NewLivenessProber(
 // ProbeLiveness implements Prober.
 func (l *livenessProber) ProbeLiveness(ctx context.Context, region string) error {
 	// If region liveness is disabled then nothing to do.
-	if !RegionLivenessEnabled.Get(&l.settings.SV) {
+	regionLivenessEnabled, tableTimeout := l.GetProbeTimeout()
+	if !regionLivenessEnabled {
 		return nil
 	}
 	const probeQuery = `
 SELECT count(*) FROM system.sql_instances WHERE crdb_region = $1::system.crdb_internal_region
 `
-	err := timeutil.RunWithTimeout(ctx, "probe-liveness", probeLivenessTimeout,
+	err := timeutil.RunWithTimeout(ctx, "probe-liveness", tableTimeout,
 		func(ctx context.Context) error {
 			return l.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
 				_, err := txn.QueryRowEx(
@@ -203,10 +201,11 @@ func (l *livenessProber) QueryLiveness(ctx context.Context, txn *kv.Txn) (LiveRe
 	return regionStatus, nil
 }
 
-// GetTableTimeout gets maximum timeout waiting on a table before issuing
+// GetProbeTimeout gets maximum timeout waiting on a table before issuing
 // liveness queries.
-func (l *livenessProber) GetTableTimeout() (bool, time.Duration) {
-	return RegionLivenessEnabled.Get(&l.settings.SV), probeLivenessTimeout
+func (l *livenessProber) GetProbeTimeout() (bool, time.Duration) {
+	return RegionLivenessEnabled.Get(&l.settings.SV),
+		RegionLivenessProbeTimeout.Get(&l.settings.SV)
 }
 
 // IsQueryTimeoutErr determines if a query timeout error was hit, specifically

